---
title: "Transcriptome Demo"
author: "Tricia"
date: "`r Sys.Date()`"
output: pdf_document
---

# Load required packages
```{r message=FALSE}
library(ballgown)
library(RColorBrewer)
library(genefilter)
library(dplyr)
library(devtools)
library(BiocManager)
```

# This code creates a table with two columns (ids) and stage shows the stage of each sample
#out put is 4 observations with 2 variables 
```{r}
pheno_data <- data.frame(ids = c("WT01", "WT02", "FurC01", "FurC02"),
                         stage = c("Wild-type", "Wild-type", "Fur mutant", "Fur mutant"))
```

# create Ballgown object and check transcript number
#output: ballgown instance with 2330 transcripts and 4 samples
```{r message=FALSE}
samples_c <- paste('ballgown', pheno_data$ids, sep = '/')
bg <- ballgown(samples = samples_c, meas='all', pData = pheno_data)
bg
```

# This code is filtering out the bg in the data and keeping the rows that have high levels of gene expression
#we are making a new variable called bg_filt : ballgown instance with 2054 transcripts and 4 samples
```{r}
bg_filt = subset(bg,"rowVars(texpr(bg)) >1",genomesubset=TRUE)
bg_filt
```

# create a table of transcripts
```{r}
results_transcripts<- stattest(bg_filt, feature = "transcript", covariate = "stage",
getFC = TRUE, meas = "FPKM")
results_transcripts<-data.frame(geneNames=geneNames(bg_filt),
transcriptNames=transcriptNames(bg_filt), results_transcripts)
```

# Picking a transcript to examine more closely
#I picked gene-HT085_RS00005 (dnaA)
```{r}
results_transcripts[results_transcripts$transcriptNames == "gene-HT085_RS00005", ]
```
## what information are you given about this transcript?
# We are given the gene name (dnaA) the transcript name (gene-HT085_RS00005) the feature (transcript) the id (1) the fc (0.7425191), the pval (0.3016928), and hte qval (0.6669607)

# Filtering the data called results_transcripts, and only picking rows that have a pval less than 0.05 meaning its significant
#Dim = calculating the dimentions and printing the resutls
# output: [1] 224   7
```{r}
sigdiff <- results_transcripts %>% filter(pval<0.05)
dim(sigdiff)
```

# organize the table BY: the p value (smaller to larger) and by fold change (descending absolute fold change) 
```{r}
o = order(sigdiff[,"pval"], -abs(sigdiff[,"fc"]), decreasing=FALSE)
output = sigdiff[o,c("geneNames","transcriptNames", "id","fc","pval","qval")]
write.table(output, file="SigDiff.txt", sep="\t", row.names=FALSE, quote=FALSE)
head(output)
```

# load gene names
```{r}
bg_table = texpr(bg_filt, 'all')
bg_gene_names = unique(bg_table[, 9:10])
```

# pull out gene expression data and visualize
```{r}
gene_expression = as.data.frame(gexpr(bg_filt))
head(gene_expression)
```

# This code gives an output of differnet coloumn names, and outputs [1] 1627    4
#this code is naming the columns to match the samples from before, and displaying the dataset with dimensions 
```{r}
colnames(gene_expression) <- c("WT01", "WT02", "FurC01", "FurC02")
head(gene_expression)
dim(gene_expression)
```

# load the transcript to gene table and determine the number of transcripts and unique genes
#output: Rconsole: [1] 2330, [1] 1744 and a table with t_id and g_id with 6 rows 

```{r}
transcript_gene_table = indexes(bg)$t2g
head(transcript_gene_table)
length(row.names(transcript_gene_table))
length(unique(transcript_gene_table[,"g_id"]))
```

# plot the number of transcripts per gene
```{r}
counts=table(transcript_gene_table[,"g_id"])
c_one = length(which(counts == 1))
c_more_than_one = length(which(counts > 1))
c_max = max(counts)
hist(counts, breaks=50, col="bisque4", xlab="Transcripts per gene",
main="Distribution of transcript count per gene")
legend_text = c(paste("Genes with one transcript =", c_one),
paste("Genes with more than one transcript =", c_more_than_one),
paste("Max transcripts for single gene = ", c_max))
legend("topright", legend_text, lty=NULL)
```


#This graph shows that majoirity of the genes in this set have 0-1 transcripts per gene, meaning the dataset has a lower number of transcripts as a whole. This graph gives insight into distribution of data and shows if there is high diversity or not. 

#Testing how similar the WT replicates are to each other 
```{r}
x = gene_expression[,"WT01"] 
y = gene_expression[,"WT02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="blue", cex=0.25,
xlab="FPKM (Wild-type, Replicate 1)", ylab="FPKM (Wild-type, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "hotpink")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

#Testing how similar the Mutant replicates are to each other 
```{r}
x = gene_expression[,"FurC01"]
y = gene_expression[,"FurC02"]
min_nonzero=1
plot(x=log2(x+min_nonzero), y=log2(y+min_nonzero), pch=16, col="darkgreen", cex=0.25,
xlab="FPKM (Fur mutant, Replicate 1)", ylab="FPKM (Fur mutant, Replicate 2)",
main="Comparison of expression values for a pair of replicates")
abline(a=0,b=1, col = "purple")
rs=cor(x,y)^2
legend("topleft", paste("R squared = ", round(rs, digits=3), sep=""), lwd=1, col="black")
```

#If the data sets are similar, which they WT is not, but mutant is, it means that there is/ is not much variablity between the replicates. WT has very low correlation compared to mutant Fur, shows that the mutant data set is more reproducable.  

# creating a plot of differential gene expression between the conditions
```{r}
results_genes = stattest(bg_filt, feature="gene", covariate="stage", getFC=TRUE, meas="FPKM")
results_genes = merge(results_genes,bg_gene_names,by.x=c("id"),by.y=c("gene_id"))
sig=which(results_genes$pval<0.05)
results_genes[,"de"] = log2(results_genes[,"fc"])
hist(results_genes[sig,"de"], breaks=50, col="hotpink",
xlab="log2(Fold change) Wild-type vs Fur mutant",
main="Distribution of differential expression values")
abline(v=-2, col="black", lwd=2, lty=2)
abline(v=2, col="black", lwd=2, lty=2)
legend("topleft", "Fold-change > 4", lwd=2, lty=2)
```

## The two dotted black lines show the thresholds for downregulated and upregulated genes. A negative fold change (left side of graph) means downregulation of WT vs FurC. In this graph there is higher points of freq on the right (positive) side meaning, there is a larger amount of upregulated genes in FurC, meaning there are some genetic differences between the two. 

#Ploting the total gene expression highlighting differentially expressed genes
```{r}
gene_expression[,"WT"]=apply(gene_expression[,c(1:2)], 1, mean)
gene_expression[,"FurC"]=apply(gene_expression[,c(3:4)], 1, mean)
x=log2(gene_expression[,"WT"]+min_nonzero)
y=log2(gene_expression[,"FurC"]+min_nonzero)
plot(x=x, y=y, pch=16, cex=0.25, xlab="Wild-type FPKM (log2)", ylab="Fur mutant FPKM (log2)",
main="Wild-type vs Fur mutant FPKMs")
abline(a=0, b=1)
xsig=x[sig]
ysig=y[sig]
points(x=xsig, y=ysig, col="seagreen", pch=16, cex=0.5)
legend("topleft", "Significant", col="seagreen", pch=16)
```

#Making a table of FPKM values
```{r}
fpkm = texpr(bg_filt,meas="FPKM")
```

#Choosing a gene to determine individual expression I picked 2, ggene-HT085_RS00010, dnaN
```{r}
ballgown::transcriptNames(bg_filt)[2]
ballgown::geneNames(bg_filt)[2]
```
# transform to log2 (change to 2)
```{r}
transformed_fpkm <- log2(fpkm[2, ] + 1)
```

# make sure values are properly coded as numbers
```{r}
numeric_stages <- as.numeric(factor(pheno_data$stage))

jittered_stages <- jitter(numeric_stages)
```

# plot expression of individual
```{r}
boxplot(transformed_fpkm ~ pheno_data$stage,
        main=paste(ballgown::geneNames(bg_filt)[2], ' : ', ballgown::transcriptNames(bg_filt)[2]),
        xlab="Stage",
        ylab="log2(FPKM+1)",
        col=c("lightblue", "salmon"),
        border="darkblue")

points(transformed_fpkm ~ jittered_stages, 
       pch=21, col="blue", bg="lightblue", cex=1.2)
```

## interpret the above figure
#this figure shows a visual for distribution of gene expression across diff stages in box-plot form for dnaN. FurC is important part of regualating expression of dnaN since it is in charge of creating the beta clamp, so mutated furC makes sense to have less FurC expression. 